### PR TITLE
Refresh and enable the ComInterop code in PowerShell

### LINKED ISSUE #12842: Refresh and enable the 'ComBinder' code in PowerShell
The support for `dynamic` and `ComBinder` was added by https://github.com/dotnet/runtime/pull/33060 and shipped in .NET 5 preview.2.
We should refresh the code under `src\System.Management.Automation\engine\ComInterop` with the code from [dotnet/runtime](https://github.com/dotnet/runtime/tree/ac219a0b106b380c7946eb2c6c1f4b182b39e43a/src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop) and enable `ComBnder` code path in our binders.

`ComBinder` will significantly speed up COM interop in PowerShell. See the reference issue: https://github.com/dotnet/runtime/issues/36944

### PR DESCRIPTION
**Pull request by daxian-dbw**
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

Fix #12842

Refresh and enable the `ComInterop` code in PowerShell to get significant perf improvement in COM related operations.

## Perf Comparison

Here is the perf difference shown by running the real world script provided in https://github.com/PowerShell/PowerShell/issues/12842#issuecomment-636313386
It's 20x faster with this PR.

PowerShell 7.0.3
```
Started script

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Could not find Timezones.csv in C:\arena\tmp\comInterop\testing.
Script will continue, ID will be used instead of time-zone name.
Config system global line found.
Section done.
Config system interface line found.
Section done.
Config router static line found.
Section done.
Creating Table of Contents
Writing Excelfile C:\arena\tmp\comInterop\testing\PerfTest_20200526_0942.xls
Script done in 2 Minute(s) and 6 Second(s).
```

Private build with this PR
```
Started script

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              Could not find Timezones.csv in C:\arena\tmp\legacy-comInterop-code\real-world-test-script.
Script will continue, ID will be used instead of time-zone name.
Config system global line found.
Section done.
Config system interface line found.
Section done.
Config router static line found.
Section done.
Creating Table of Contents
Writing Excelfile C:\arena\tmp\legacy-comInterop-code\real-world-test-script\PerfTest_20200526_0942.xls
Script done in 0 Minute(s) and 6 Second(s).
```

## ComInterop Context

> **NOTE: _This comes from the content in `engine/ComInterop/README.md`, please review for the detailed information._**

The ComInterop code shipped in PowerShell comes from [dotnet/runtime](https://github.com/dotnet/runtime) with _a considerable amount of refactoring work_ to make it work properly with PowerShell.

There are 3 sources of the ComInterop code as our references:

1. [The .NET Framework version](https://github.com/IronLanguages/main/tree/ipy-2.7-maint/Runtime/Microsoft.Dynamic/ComInterop).
   The code was archived in 2012.

2. [The .NET 5.0 version](https://github.com/dotnet/runtime/tree/master/src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop).
   It was merged into .NET 5.0 in May 2020 through the PR [dotnet/runtime#33060](https://github.com/dotnet/runtime/pull/33060).
   It's based on the .NET Framework version code with quite amount of refactoring work.

3. [The legacy ComInterop code from Windows PowerShell](https://github.com/PowerShell/PowerShell/tree/v7.0.0/src/System.Management.Automation/engine/ComInterop).
   The legacy code has always been in the repository, but it was excluded from compilation.
   It was based on the .NET Framework version code with a considerable amount of refactoring work to make it work properly with Windows PowerShell.

### Code Refreshing

The ComInterop code was refreshed and enabled in compilation in PowerShell 7.1, August 2020.
It was done manually by:

- A careful **three-way comparison** across all the code sources listed above:
  - Compare (3) to (1) to get the PowerShell specific changes
  - Compare (2) to (1) to get the .NET 5.0 changes

- Applying the PowerShell specific changes from (3) to (2), with necessary further refactoring

- A careful review of the new PowerShell specific changes applied to (2), again using the three-way comparison:
  - Compare the refreshed ComInterop code with (3) to get the differences
  - Analyze the differences to make sure they are either the pure .NET 5.0 changes over the .NET Framework code,
    or the refactoring changes that is necessary to apply the PowerShell specific changes.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs [hunk 1]
```diff
@@ -78,8 +68,8 @@ internal bool Implements(string itfName, bool isSourceItf)
         {
             if (isSourceItf)
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs:69]: brackets?

File: src/System.Management.Automation/engine/ComInterop/ConvertibleArgBuilder.cs [hunk 2]
```diff
@@ -1,36 +1,22 @@
...
         {
             return Helpers.Convert(parameter, typeof(IConvertible));
         }
 
         internal override Expression MarshalToRef(Expression parameter)
         {
-            // we are not supporting convertible InOut
-            throw Assert.Unreachable;
+            //we are not supporting convertible InOut
+            throw new NotSupportedException();
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/ComInterop/ConvertibleArgBuilder.cs:19]: PSNotSupportedException?

File: src/System.Management.Automation/engine/ComInterop/InteropServices/ComEventsMethod.cs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        /// correctly but one known case is when the expected signature is 'ref Enum' - in this case
+        /// reflection by design does not do the coercion. Since we need to be compatible with COM interop
+        /// handling of this scenario - we are pre-processing delegate's signature by looking for 'ref enums'
+        /// and cache the types required for such coercion.
+        /// </summary>
+        public class DelegateWrapper
+        {
+            private bool _once;
+            private int _expectedParamsCount;
+            private Type?[]? _cachedTargetTypes;
```
Comment (PaulHigin) [inline src/System.Management.Automation/engine/ComInterop/InteropServices/ComEventsMethod.cs:31]: I have not seen this syntax before...

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/engine/ComInterop/BoundDispEvent.cs [hunk 4]
```diff
@@ -78,9 +73,10 @@ private static void VerifyHandler(object handler)
         /// <returns>The original event with handler added.</returns>
         private object InPlaceAdd(object handler)
         {
+            Requires.NotNull(handler, nameof(handler));
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/BoundDispEvent.cs:76]: This is from the .NET 5.0 code. They added `Requires.NotNull` in lots of places.

File: src/System.Management.Automation/CoreCLR/CorePsStub.cs [hunk 5]
```diff
@@ -590,6 +343,89 @@ public void Dispose()
...
+namespace System.Management.Automation.ComInterop
+{
+    using System.Dynamic;
+    using System.Runtime.InteropServices;
+
+    /// <summary>
+    /// Provides helper methods to bind COM objects dynamically.
+    /// </summary>
+    /// <remarks>
+    /// COM is not supported in core powershell. So this is a stub type.
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/CoreCLR/CorePsStub.cs:355]: Good catch! Will update the comment.

File: src/System.Management.Automation/engine/ComInterop/ComBinder.cs [hunk 6]
```diff
@@ -239,7 +224,7 @@ public static bool TryConvert(ConvertBinder binder, DynamicMetaObject instance,
                         ),
                         BindingRestrictions.GetExpressionRestriction(
                             Expression.Call(
-                                typeof(ComObject).GetMethod("IsComObject", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.NonPublic),
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/ComBinder.cs:242]: Yes, it's intentional. The `ComBinder.IsComObject` is defined at the top of this file, as a public method:
```
public static bool IsComObject(object value)
{
    return value != null && Marshal.IsComObject(value);
}
```

File: src/System.Management.Automation/engine/ComInterop/ComBinder.cs [hunk 7]
```diff
@@ -252,26 +237,17 @@ public static bool TryConvert(ConvertBinder binder, DynamicMetaObject instance,
...
-        }
-
         /// <summary>
         /// Gets the member names of the data-like members associated with the object.
         /// This function can operate only with objects for which <see cref="IsComObject"/> returns true.
         /// </summary>
         /// <param name="value">The object for which member names are requested.</param>
         /// <returns>The collection of member names.</returns>
-        [SuppressMessage("Microsoft.Performance", "CA1811:AvoidUncalledPrivateCode")]
         internal static IList<string> GetDynamicDataMemberNames(object value)
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/ComBinder.cs:246]: It's not used currently, but it's better to keep it so we have minimal diffs comparing with the .NET 5 code.

File: src/System.Management.Automation/engine/ComInterop/ComEventSinksContainer.cs [hunk 8]
```diff
@@ -63,9 +59,9 @@ public void Dispose()
 
         private void DisposeAll()
         {
-            foreach (ComEventSink sink in this)
+            foreach (ComEventsSink sink in this)
             {
-                sink.Dispose();
+                ComEventsSink.RemoveAll(sink);
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/ComEventSinksContainer.cs:64]: I believe so, it eventually calls to [sink.Unadvise()](https://github.com/PowerShell/PowerShell/pull/13304/files/dde98ed8dc6dcea81a54280381b69c717821721a#diff-6ad404d55b0ebcfefc7233b8f5797b93R263-R284) where `Marshal.ReleaseComObject` is called to release some COM objects. The `ComEventsSink.cs` is one of the biggest refactoring work of this ComBinder done in .NET 5.

File: src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs:69]: It's a copy of the .NET code. I prefer to stick to the style used in the code, so we can keep the diffs minimal when comparing with the .NET 5 code.

File: src/System.Management.Automation/engine/ComInterop/ConvertibleArgBuilder.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/ConvertibleArgBuilder.cs:19]: Same here. I prefer to keep the code as is, so the diffs between the code here and in dotnet/runtime is minimal.

File: src/System.Management.Automation/engine/ComInterop/InteropServices/ComEventsMethod.cs (hunk 3 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/InteropServices/ComEventsMethod.cs:31]: This is the nullable reference type syntax. It means `the type array could be null` and `elements in the type array` could be null.

File: src/System.Management.Automation/CoreCLR/CorePsStub.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/CoreCLR/CorePsStub.cs]: Done.

**Comment by iSazonov** [comment]
We borrow .Net code from several places such as Interpreter and ComInterop. Make sense to ask .Net team to enhance .Net API so that we can use _public_ code without copy-paste?

**Comment by daxian-dbw** [comment]
We had the conversation with the .NET team and the conclusion so far is to keep it as is today, due to 2 main reasons:
1. Both the expression interpreter and the COM binder code are archived components in .NET, there is no activity in those components (except for the needed changes to get them in .NET Core/5.0).
2. It's not a sheer copy of the .NET code. There are additional PowerShell specific changes to those code in PowerShell repo. Take [this commit](https://github.com/PowerShell/PowerShell/pull/13304/commits/373116c613f6b533c05d4120c241cb23eaf6fb3e) as an example to see the PowerShell specific changes to COM binder on top of the .NET Core 5.0 code.

**Comment by iSazonov** [comment]
> Both the expression interpreter and the COM binder code are archived components in .NET

Clear about COM binder but can you share more information about Interpreter? The "archived" sounds like one step before obsolete and two step before removing. Is C# scripting future? :-)

File: src/System.Management.Automation/engine/ComInterop/BoundDispEvent.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (PaulHigin) [reply src/System.Management.Automation/engine/ComInterop/BoundDispEvent.cs:76]: Yes, but you can reduce code duplication by adding it to VerifyHandler(), correct?

File: src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs (hunk 1 above)
Comment (PaulHigin) [reply src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs:69]: Can you add a comment to that effect?  Chances are a code clean up effort will change it.

**Review by PaulHigin — APPROVED** [review]
It sounds like this is more or less .NET 5 code, and we don't want to modify it.  Please include comments to make this clear, so that later maintenance efforts don't modify it.

File: src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/ComInterop/ComTypeClassDesc.cs:69]: Yes, I will add a comment in the `engine/ComInterop/README.md` about this.

**Comment by daxian-dbw** [comment]
> Clear about COM binder but can you share more information about Interpreter?

To the expression interpreter, the additional changes includes `UpdatePositionInstruction` and the changes to `LightLambda` to JIT compile powershell loop and script block when the threshold is met, and some more that I cannot recall now.

> The "archived" sounds like one step before obsolete and two step before removing. Is C# scripting future? :-)

`archived` means the code is relatively pretty stable and there is no new feature planned for the component. It doesn't imply `obsolete` -- bringing back COM binder in .NET 5.0 is a clear evidence that the com binder is supported.

For the expression interpreter, `System.Linq.Expression` and `Microsoft.CSharp` are where the code lives, and they are now the archived components. But those 2 libraries are obviously important building blocks for many other components.

Here is the area doc if you are interested: https://github.com/dotnet/runtime/blob/master/docs/area-owners.md

**Pull request merged**